Add support for extension VK_EXT_headless_surface.

billhollings opened the pull request:
- Consolidate info about `CAMetalLayer` and headless in `MVKSurface`.
- `MVKSwapchainImage` remove `getCAMetalDrawable()` and focus on abstracting `getMTLTexture()`.
- `MVKPresentableSwapchainImage::getCAMetalDrawable()` return `nil` if headless.
- Add `MVKPresentableSwapchainImage::_mtlTextureHeadless` to support a fixed `MTLTexture` that is not retrieved from a `CAMetalDrawable`.
- `MVKPresentableSwapchainImage` refactor signalling semaphores and fences.
- `MVKPresentableSwapchainImage` don't lock when signalling semaphores and fences.
- If no present occurs, `actualPresentTime` will be zero. Set it to current time, instead of to `desiredPresentTime`, since it's more accurate.
- Rework timestamps:
  - Remove `_mvkTimestampBase` so `mvkGetTimestamp()` is equal to `mach_absolute_time()`, which is used in presentation timing.
  - Add `mvkGetRuntimeNanoseconds()`.
  - Rename `mvkGetAbsoluteTime()` to `mvkGetContinuousNanoseconds()`.
  - Remove `mvkGetTimestampPeriod()` as unused.
- `MVKSemaphoreMTLEvent::encodeDeferredSignal` remove redundant `nil` test (unrelated).
- Fix swapchain and surface bugs when windowing system  is accessed from off the main thread (unrelated).
- Log warning when deprecated functions `vkCreateMacOSSurfaceMVK()` 
  or `vkCreateIOSSurfaceMVK()` are used (unrelated).
- Remove documentation for `visionos`, as support is not ready (unrelated).


Fixes #2049.
Fixes #2085.

Passes all related CTS WSI tests.

billhollings commented:
@cdavis5e 

The (now renamed) `mvkGetContinuousNanoseconds()` was introduced for `VK_EXT_calibrated_timestamps`. Is there a reason why that time must be continuous (`mach_continuous_time()`)? Or can it be replaced with `mvkGetRuntimeNanoseconds()`?

cdavis5e commented:
> Is there a reason why that time must be continuous (`mach_continuous_time()`)?

Yes. This is the function that's used for the `CLOCK_MONOTONIC_RAW` domain. `VK_EXT_calibrated_timestamps` doesn't have `CLOCK_UPTIME` or `CLOCK_UPTIME_RAW` domains. See the comment starting on [line 1182 of `MVKDevice.mm`](https://github.com/KhronosGroup/MoltenVK/blob/8820c53666644ecbcc66e88f7c9d26257f9e66aa/MoltenVK/MoltenVK/GPUObjects/MVKDevice.mm#L1182).

cdavis5e commented on MoltenVK/MoltenVK/GPUObjects/MVKImage.mm:
```diff
@@ -1408,14 +1405,25 @@ static void signalAndUnmarkAsTracked(const MVKSwapchainSignaler& signaler) {
+	// If the presentation time is not available, use the current nanosecond runtime clock,
+	// which should be reasonably accurate (sub-ms) to the presentation time. The presentation
+	// time will not be available if the presentation did not actually happen, such as when
+	// running headless, or on test harnesses that is not attached to the windowing system.
```
"test harnesses that is"??
```suggestion
	// running headless, or on test harnesses that are not attached to the windowing system.
```

billhollings replied:
Thanks. Fixed.

cdavis5e commented on MoltenVK/MoltenVK/GPUObjects/MVKSwapchain.mm:
```diff
@@ -618,14 +610,19 @@ static CALayerContentsGravity getCALayerContentsGravity(VkSwapchainPresentScalin
-	if ([mtlLayer.screenMVK respondsToSelector:@selector(localizedName)]) {
-		screenName = mtlLayer.screenMVK.localizedName;
-	}
+		if ([mtlLayer.screenMVK respondsToSelector:@selector(localizedName)]) {
```
And this call *definitely* needs to happen on the main thread, because that's the one that triggered the crash the reporter saw.

billhollings replied:
Ditto.

billhollings commented on MoltenVK/MoltenVK/GPUObjects/MVKSwapchain.mm:
```diff
@@ -187,30 +190,29 @@
-#if MVK_VISIONOS
-	// TODO: See if this can be obtained from OS instead
-	NSInteger framesPerSecond = 90;
+	auto* screen = getCAMetalLayer().screenMVK;		// Will be nil if headless
```
Thanks for noticing. Fixed.

billhollings commented:
Thanks for noticing the connection to #2085.

While I was in there, I made a couple of additional changes about windowing systems. And while I was doing _that_, I also removed the documentation for `visionos` support, as it's really not properly supported and doesn't build correctly. Also added a deprecation warning for the old macOS and iOS surface creation extensions.
